flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/AlgoOperator.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.flink.ml.common;
+
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+import org.apache.flink.ml.api.misc.param.Params;
+import org.apache.flink.ml.api.misc.param.WithParams;
+import org.apache.flink.ml.batchoperator.source.TableSourceBatchOp;
+import org.apache.flink.ml.params.shared.HasMLEnvironmentId;
+import org.apache.flink.ml.streamoperator.source.TableSourceStreamOp;
+import org.apache.flink.table.api.Table;
+import org.apache.flink.table.api.TableSchema;
+import org.apache.flink.table.api.internal.TableImpl;
+import org.apache.flink.table.api.java.StreamTableEnvironment;
+
+import java.io.Serializable;
+
+/**
+ * Base class for algorithm operators.
+ *
+ * <p>Hold a Table as its output.
+ *
+ * @param <T> The class type of the {@link AlgoOperator} implementation itself
+ */
+public abstract class AlgoOperator<T extends AlgoOperator <T>>
+	implements WithParams<T>, HasMLEnvironmentId<T>, Serializable {
+
+	/**
+	 * Params for algorithms.
+	 */
+	private Params params;
+
+	/**
+	 * The table held by operator.
+	 */
+	private Table output = null;
+
+	/**
+	 * The side outputs of operator that be similar to the stream's side outputs.
+	 */
+	private Table[] sideOutputs = null;
+
+	/**
+	 * Construct the operator with empty Params.
+	 */
+	protected AlgoOperator() {
```

Comment:
The Table is the output of the operator. It should be set at linkFrom and call getOutput to get the output for the next linked operator in most of the scenes. So the AlgoOperator() and the AlgoOperator(Params params) should be exist.
The TableSourceBatchOp and the TableSourceStreamOp is the only two cases that construct the Operator using the table. These operators are the special cases and need not add a new constructor in the base class for these cases.

Reply:
I think the main point here is that since the `public AlgoOperator()`  API is not used/tested. we can add it in from later PRs, correct me if I were wrong @ex00 ? 
If that's the case then yes I do agree that any new APIs added which are not tested should be part of latter PRs.

Reply:
yes @walterddr, you are right. I don't see how using this constructor here.
Also, personally for me, logic in parent class, when using `output` field and it is not marked 'as required', is confusing a little bit.

Reply:
In later PRs, these APIs will be used and tested.

For how using this constructor, I'd like to give an example. 
SplitBatchOp is widely used in ML data pre-preprossing, which splits one dataset input 2 dataset: training set and validation set. It is very convenient for us to write code like this:
new SplitBatchOp().setSplitRatio(0.9)

For output field. The AlgoOperator may have one or more result tables, in most cases, it has only one result. The output is the main operation result table, and the other results are kept in the sideOutputs.
For example, 2 AlgoOperators: AlgoA and AlgoB, AlgoB takes the AlgoA’s results as its inputs, we can write the code like this:
AlgoB.linkFrom(AlgoA)
AlgoA.getOutput() provides the main result of AlgoA, and AlgoA.getSideOutputs() provides the other results of AlgoA. AlgoB will take the AlgoA’s results as its inputs by calling AlgoA.getOutput() and AlgoA.getSideOutputs().

Reply:
>new SplitBatchOp().setSplitRatio(0.9)

How will it be apply to data source and how will define output result? Could you explain more please?

Reply:
Thanks, we have the `setOutput` and can set the output in any method of the operator

Reply:
is it mean what it will be look like `new SplitBatchOp().setSplitRatio(0.9).setOutput(table)` ?
if it is right assumption, that mean that who will write code need to know that need to call `setOutput`. And if someone will doing something from you example, easier to lost this call, and exception will be throw in run time only
```
AlgoA = new SplitBatchOp().setSplitRatio(0.9)
AlgoB.linkFrom(AlgoA)
AlgoA.getSchema()  // NPE: output is null
```
But if we define that `SplitBatchOp(table)` and developer don't thing about additional method calls and will get error on compile time in wrong case.

Reply:
Yes, there must be one or more data sources. We have defined some common *SourceBatchOp to get the source data conveniently, and from the sourceBatchOp, we can link the algorithm operations.
Here is a whole example:
```
CsvSourceBatchOp algoA =
	new CsvSourceBatchOp()
		.setFilePath("http://alink-dataset.cn-hangzhou.oss.aliyun-inc.com/csv/iris.csv")
		.setSchemaStr("s_length double, s_width double, p_length double, p_width double, category string");

System.out.println("schema of algoA:");
System.out.println(algoA.getSchema());

SplitBatchOp algoB = new SplitBatchOp().setFraction(0.8);

algoB.linkFrom(algoA);

System.out.println("schema of algoB's main result:");
System.out.println(algoB.getSchema());

System.out.println("schema of algoB's side output result:");
System.out.println(algoB.getSideOutput().getSchema());
```

And the print results:
```
schema of algoA:
root
 |-- s_length: Double
 |-- s_width: Double
 |-- p_length: Double
 |-- p_width: Double
 |-- category: String

schema of algoB's main result:
root
 |-- s_length: Double
 |-- s_width: Double
 |-- p_length: Double
 |-- p_width: Double
 |-- category: String

schema of algoB's side output result:
root
 |-- s_length: Double
 |-- s_width: Double
 |-- p_length: Double
 |-- p_width: Double
 |-- category: String
```

Reply:
now I see, thanks for the explanation